hail/python/hail/expr/expressions/base_expression.py

```diff
@@ -339,16 +339,16 @@ def describe(self, handler=print):
         handler(s)
 
     def __lt__(self, other):
-        raise NotImplementedError("'<' comparison with expression of type {}".format(str(self._type)))
+        return self._bin_op("<", other, hl.tbool)
```

Comment:
We need to check types -- it's possible to generate invalid computations which the backend won't like:

```
hl.literal('abc') < hl.struct(foo=1)
```

The solution is to copy and modify this bit from `__eq__`:

```
        other = to_expr(other)
        left, right, success = unify_exprs(self, other)
        if not success:
            raise TypeError(f"Invalid '==' comparison, cannot compare expressions "
                            f"of type '{self.dtype}' and '{other.dtype}'")
        return left._bin_op("==", right, tbool)
```

not only does this check types, it also unifies types where possible, e.g. `tuple<bool, float32>` and `tuple<int32, int32>` become comparable by unifying both to a `tuple<int32, float32>`.

Reply:
I see. I will submit the change. Thanks!